### PR TITLE
symbols: go-build only builds target binary

### PR DESCRIPTION
**Pull request by keegancsmith**
Previously go-build.sh just built the full symbols docker image and then copied out the symbols binary. This did a lot of extra work, including building ctags. go-build.sh is called as part of building the server image, one of the slowest steps on the critical path in CI. It now just runs the build target in the Dockerfile.

There are multiple improvements to the Dockerfile. This change was motivated by switching to go1.18. Before this commit when using go1.18 the server build would fail in CI due to symbol's go-build.sh. I didn't fully root cause it, but I believe it had something to do with using alpine-3.14 as the builder image which included an older version of go.

This commits moves just symbols to use go1.18. Another commit will move everything else to go1.18.

These are the other improvements made in this commit:
- ARGs come last to prevent them invalidating caches.
- build target renamed to symbols-build to be more clear in CI.
- --no-cache used in apk add.
- ENTRYPOINT uses JSON args due to hadolint complaining.
- DOCKER_BUILDKIT=1 is set to simplify running locally.

Test Plan: main-dry-run on CI will exercise building the images and will run integration tests. To test locally I ran

```shell
VERSION=test PKG=github.com/sourcegraph/sourcegraph/enterprise/cmd/symbols ./cmd/symbols/go-build.sh /tmp
```

### CODE REVIEW

File: cmd/symbols/Dockerfile [hunk 1]
```diff
@@ -6,27 +6,27 @@ USER root
...
-ARG PKG
-ENV PKG=$PKG
-
 ENV GO111MODULE on
 ENV GOARCH amd64
 ENV GOOS linux
 ENV CGO_ENABLED 1
 
-RUN apk add go gcc g++
+RUN apk add --no-cache go gcc g++
```
Comment (bobheadxi) [inline cmd/symbols/Dockerfile:18]: If we're coming from an `golang:1.18.1` image do we still need to install go here?

File: cmd/symbols/go-build.sh [hunk 2]
```diff
@@ -2,12 +2,25 @@
...
 
-docker cp "$(docker create --rm "$IMAGE")":/usr/local/bin/symbols "$OUTPUT/symbols"
+# Required due to use of RUN --mount=type=cache in Dockerfile.
+export DOCKER_BUILDKIT=1
+
+docker build -f cmd/symbols/Dockerfile -t symbols-build "$(pwd)" \
+  --target=symbols-build \
+  --progress=plain \
+  --build-arg VERSION \
+  --build-arg PKG="${PKG:-github.com/sourcegraph/sourcegraph/cmd/symbols}"
```
Comment (bobheadxi) [inline cmd/symbols/go-build.sh:24]: Why does `PKG` need to be parameterized here? (seems like it wasn't before)

**Review by bobheadxi — APPROVED** [review]
LGTM overall! Thanks for the cleanup here 😁

File: cmd/symbols/Dockerfile (hunk 1 above)
Comment (keegancsmith) [reply cmd/symbols/Dockerfile]: you are right, tested and it doesn't need it :) pushed a commit.

File: cmd/symbols/go-build.sh (hunk 2 above)
Comment (keegancsmith) [reply cmd/symbols/go-build.sh:24]: It was there before, see previously we called `build.sh` which has the same build arg. In practice we always have PKG set to `github.com/sourcegraph/sourcegraph/enterprise/cmd/symbols` in CI. IE we actually call `enterprise/cmd/symbols/go-build.sh` when building the enterprise `server` docker image. Not confusing at all :P

**Pull request merged**